src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,185 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private Object[] array = new Object[DEFAULT_CAPACITY];
+    private int size = 0;
+
     @Override
     public void add(T value) {
-
+        if (array.length > size) {
+            addElement(value);
+        } else {
+            grow();
+            addElement(value);
+        }
+        size++;
     }
 
     @Override
-    public void add(T value, int index) {
-
+    public void add(T value, int index) throws ArrayListIndexOutOfBoundsException {
+        checkIfIndexNegative(index);
+        checkIfIndexPresent(index);
+        if (array.length > size) {
+            addElement(value, index);
+        } else {
+            grow();
+            addElement(value, index);
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        int newSize = size + list.size();
+        if (newSize < array.length) {
+            copyArray(list);
+        } else {
+            checkIfArrayShouldGrow(list);
+            copyArray(list);
+        }
 
+        size += list.size();
     }
 
     @Override
-    public T get(int index) {
-        return null;
+    public T get(int index) throws ArrayListIndexOutOfBoundsException {
+        checkIfIndexNegative(index);
+        checkIfPositionExistent(index);
+        T result = (T) array[index];
+        return result;
     }
 
     @Override
-    public void set(T value, int index) {
-
+    public void set(T value, int index) throws ArrayListIndexOutOfBoundsException {
+        checkIfIndexNegative(index);
+        checkIfPositionExistent(index);
+        array[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIfIndexNegative(index);
+        checkIfPositionExistent(index);
+        Object[] result = Arrays.copyOf(array, array.length - 1);
+        T removedItem = null;
+        removedItem = (T) array[index];
+        System.arraycopy(array, index + 1, result, index, array.length - index - 1);
+        array = result;
+        size--;
+        return removedItem;
     }
 
     @Override
-    public T remove(T element) {
+    public T remove(T element) throws NoSuchElementException {
+        boolean flag = false;
+
+        for (int i = 0; i < size; i++) {
```

Comment:
The current implementation for 'remove(T element)' is inefficient because it involves multiple array copying operations. Consider optimizing the method to avoid unnecessary array copies.